RHDEVDOCS-5238: pipelines-as-code configuration and token scoping (#60197)
openshift/openshift-docs: modules/op-scoping-github-token.adoc

// This module is included in the following assembly:
//
// *cicd/pipelines/using-pipelines-as-code.adoc

:_content-type: PROCEDURE
[id="scoping-github-token_{context}"]
= Scoping the GitHub token to additional repositories

> {pac} uses the GitHub App to generate a GitHub access token. It uses this token both for retrieving the pipeline payload from the repository and for interactions of the CI/CD process with GitHub repositories.
> 
> By default, this token is scoped only to the repository where the payload comes from.
> However, in some cases, the developer team might want the token to allow control over additional repositories. For example, there might be a CI repository where the `.tekton/pr.yaml` file and source payload might be located, however the build process defined in `pr.yaml` might fetch tasks from a separate private CD repository.

Reviewer: ```suggestion
{pac} uses the GitHub app to generate a GitHub access token. This token is used to retrieve the pipeline payload from the repository, and to allow CI/CD processes to interact with GitHub repositories. By default, this token is scoped only to the repository where the payload comes from.

In some cases, you might want the token to have access to additional repositories. For example, there might be a CI repository where the `.tekton/pr.yaml` file and source payload are located, but the build process defined in `pr.yaml` needs to fetch tasks from a separate private CD repository.
```

suggestions

Author: I don't think adding passive voice is good, but did some rewriting.

You can extend the scope of the GitHub token in two ways:

> * _Global configuration_: extend the GitHub token to a list of repositories in different namespaces. You must have administrative permissions to set this configuration.
> * _Repository level configuration_: extend the GitHub token to a list of repositories that exist in the same namespace as the original repository. You do not require administrative permissions to set this configuration.

Reviewer: ```suggestion
* _Global configuration_: You can extend the GitHub token to a list of repositories in different namespaces. You must have administrative permissions to set this configuration.
* _Repository level configuration_: You can extend the GitHub token to a list of repositories that exist in the same namespace as the original repository. You do not require administrative permissions to set this configuration.
```

suggestion

.Procedure

> . In , set the `secret-github-app-token-scoped` key to `false`. This setting enables the scoping of the GitHub token to private and public repositories listed under the Global and Repository level configuration.

Reviewer: ```
 In ,
```
Looks like there's a word missing here?

> . To set global configuration for scoping the GitHub token, in the `TektonConfig` custom resource, under `pipelinesAsCode` and `settings`, specify the additional repositories the `secret-github-app-scope-extra-repos` key, as in the following example:

Reviewer: It seems that parameter and key are used interchangeably throughout the docs. I'm not sure if there's any guidance about whether to use key, field, parameter, etc., but I would choose one and stick with it throughout your PR for consistency.

> . To set global configuration for scoping the GitHub token, in the `TektonConfig` custom resource, under `pipelinesAsCode` and `settings`, specify the additional repositories the `secret-github-app-scope-extra-repos` key, as in the following example:

Author: While I will standardize on parameter for now, I also asked in Slack whether we have a preferred option. I will not go through the entire assembly right now though, this will await the restructure expected for the next version.

+
[source,yaml]
----
  spec:
      platforms:
        openshift:
          pipelinesAsCode:
            enable: true
            settings:
              secret-github-app-token-scoped: false
              secret-github-app-scope-extra-repos: "owner2/project2, owner3/project3"
----
+

> . To set repository level configuration for scoping the GitHub token, specify the additional repositories the `secret-github-app-scope-extra-repos` key, as in the following example:

Reviewer 2: ```suggestion
. To set repository level configuration for scoping the GitHub token, specify the additional repositories the `github_app_token_scope_repos` key, as in the following example:
```

+
[source,yaml]
----
  apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
  kind: Repository
  metadata:
    name: test
    namespace: test-repo
  spec:
    url: "https://github.com/linda/project"
    settings:
      github_app_token_scope_repos:
      - "owner/project"

>       - "owner1/project1"

Reviewer: YAML indentation should be flush to the left margin: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#formatting-1

----
+
In this example, the `Repository` custom resource is associated with the `linda/project` repository in the `test-repo` namespace. The scope of the generated GitHub token is extended to the `owner/project` and `owner1/project1` repositories, as well as the `linda/project` repository. These repositories must exist under the `test-repo` namespace.
+
[NOTE]
====
The additional repositories can be public or private, but must reside in the same namespace as the repository with which the `Repository` resource is associated.

> If any of the repositories does not exist in the namespace, the scoping of the GitHub token fails with an error message as in the following example:

Reviewer: ```suggestion
If any of the repositories do not exist in the namespace, the scoping of the GitHub token fails with an error message:
```

> [,console]

Reviewer: I haven't seen `[,console]` used before; I think we generally go with `[source,terminal]`, e.g. https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

----
failed to scope GitHub token as repo owner1/project1 does not exist in namespace test-repo
----
====

.Result

The generated GitHub token enables access to additional repositories that you configured in the global and repository level configuration, as well as the original repository from which the {pac} payload files come.

If you provide both global configuration and repository level configuration, the token is scoped to all the repositories from both configurations, as in the following example.

.`TektonConfig` custom resource
[source,yaml]
----
  spec:
      platforms:
        openshift:
          pipelinesAsCode:
            enable: true
            settings:
              secret-github-app-token-scoped: false
              secret-github-app-scope-extra-repos: "owner2/project2, owner3/project3"
----

.`Repository` custom resource
[source,yaml]
----
 apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
 kind: Repository
 metadata:
   name: test
   namespace: test-repo
 spec:
   url: "https://github.com/linda/project"
   settings:
     github_app_token_scope_repos:
     - "owner/project"
     - "owner1/project1"
----

> The GitHub token is scoped to the following repositories: `owner/project`, `owner1/project1`, `owner2/project2`, `owner3/project3`, `linda/project`.

Reviewer: I don't think that we generally have results sections, and it feels like this information about configuring both could be moved instead to under the bullet list where it's explained that you can configure global or per repository access? I feel like this info would be more useful to users at the beginning rather than here at the end.

Author: The problem is that I need to provide examples, and in the beginning examples would not work as the syntax is not explained yet. I honestly see no better solution than a results section for this, unless maybe a note is better?

// This module is included in the following assembly:
//
// *cicd/pipelines/using-pipelines-as-code.adoc

:_content-type: PROCEDURE
[id="scoping-github-token_{context}"]
= Scoping the GitHub token to additional repositories

> {pac} uses the GitHub App to generate a GitHub access token. {pac} uses this token to retrieve the pipeline payload from the repository and to enable the CI/CD processes to interact with GitHub repositories.

Reviewer: ```suggestion
{pac} uses the GitHub app to generate a GitHub access token. This token is used to retrieve the pipeline payload from the repository, and to enable the CI/CD processes to interact with GitHub repositories.
```
Unless the literal name is "Github App" I'd make app lowercase here.
Some suggestions re comma and wording, not crucial though.

Author: Can't agree with wording suggestions sorry, but doing the lowercase for app here.

> By default, this token is scoped only to the repository from which {pac} retrieves the pipeline definition. In some cases, you might want the token to have access to additional repositories. For example, there might be a CI repository where the `.tekton/pr.yaml` file and source payload might be located, but the build process defined in `pr.yaml` might fetch tasks from a separate private CD repository.

Reviewer: ```suggestion
By default, the access token is scoped only to the repository from which {pac} retrieves the pipeline definition. In some cases, you might want the token to have access to additional repositories. For example, there might be a CI repository where the `.tekton/pr.yaml` file and source payload are located, but the build process defined in `pr.yaml` fetches tasks from a separate private CD repository.
```
editing suggestions, not crucial

You can extend the scope of the GitHub token in two ways:

* _Global configuration_: You can extend the GitHub token to a list of repositories in different namespaces. You must have administrative permissions to set this configuration.

> * _Repository level configuration_: You can extend the GitHub token to a list of repositories that exist in the same namespace as the original repository. You do not require administrative permissions to set this configuration.

Reviewer: ```suggestion
* _Repository level configuration_: You can extend the GitHub token to a list of repositories that exist in the same namespace as the original repository. You do not need administrative permissions to set this configuration.
```
I'd say either "you don't need" or "are not required" rather than "you do not require", again just a suggestion for readability though

+
[source,yaml]
----
apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
kind: Repository
metadata:
  name: test
  namespace: test-repo
spec:
  url: "https://github.com/linda/project"
  settings:
    github_app_token_scope_repos:
    - "owner/project"
    - "owner1/project1"
----
+
In this example, the `Repository` custom resource is associated with the `linda/project` repository in the `test-repo` namespace. The scope of the generated GitHub token is extended to the `owner/project` and `owner1/project1` repositories, as well as the `linda/project` repository. These repositories must exist under the `test-repo` namespace.
+
[NOTE]
====
The additional repositories can be public or private, but must reside in the same namespace as the repository with which the `Repository` resource is associated.

If any of the repositories do not exist in the namespace, the scoping of the GitHub token fails with an error message:

[source,terminal]
----
failed to scope GitHub token as repo owner1/project1 does not exist in namespace test-repo
----
====

.Result

> The generated GitHub token enables access to additional repositories that you configured in the global and repository level configuration, as well as the original repository from which the {pac} payload files come.

Reviewer: ```suggestion
The generated GitHub token enables access to the additional repositories that you configured in the global and repository level configuration, as well as the original repository where the {pac} payload files are located.
```
readability suggestion - the second part of the sentence currently reads a little oddly

If you provide both global configuration and repository level configuration, the token is scoped to all the repositories from both configurations, as in the following example.

.`TektonConfig` custom resource
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  platforms:
    openshift:
      pipelinesAsCode:
        enable: true
        settings:
          secret-github-app-token-scoped: false
          secret-github-app-scope-extra-repos: "owner2/project2, owner3/project3"
----

.`Repository` custom resource
[source,yaml]
----
apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
kind: Repository
metadata:
 name: test
 namespace: test-repo
spec:
 url: "https://github.com/linda/project"
 settings:
   github_app_token_scope_repos:
   - "owner/project"
   - "owner1/project1"
----

> The GitHub token is scoped to the following repositories: `owner/project`, `owner1/project1`, `owner2/project2`, `owner3/project3`, `linda/project`.

Reviewer: ```suggestion
The GitHub token is scoped to the `owner/project`, `owner1/project1`, `owner2/project2`, `owner3/project3`, and `linda/project` respositories.
```
suggestion
